src/software/thunderscope/robot_communication.py

```diff
@@ -134,34 +116,26 @@ def run(self):
 
                 robot_primitives = dict(primitive_set.robot_primitives)
 
-            # diagnostics is running
-            if (
-                self.current_mode == RobotCommunicationMode.DIAGNOSTICS
-                or self.current_mode == RobotCommunicationMode.BOTH
-            ):
-
-                # get the manual control primitive
-                diagnostics_primitive = DirectControlPrimitive(
-                    motor_control=self.motor_control_diagnostics_buffer.get(
-                        block=False
-                    ),
-                    power_control=self.power_control_diagnostics_buffer.get(
-                        block=False
-                    ),
-                )
+            # get the manual control primitive
+            diagnostics_primitive = DirectControlPrimitive(
+                motor_control=self.motor_control_diagnostics_buffer.get(block=False),
+                power_control=self.power_control_diagnostics_buffer.get(block=False),
+            )
 
+            # diagnostics is running
+            if self.robots_connected_to_manual:
                 # for all robots connected to diagnostics, set their primitive
-                for robot_id in self.robots_connected_to_diagnostics:
+                for robot_id in self.robots_connected_to_manual:
                     robot_primitives[robot_id] = Primitive(
                         direct_control=diagnostics_primitive
                     )
 
             # sends a final stop primitive to all disconnected robots and removes them from list
             # in order to prevent robots acting on cached old primitives
-            while self.robots_to_be_disconnected:
-                robot_primitives[self.robots_to_be_disconnected.pop()] = Primitive(
-                    stop=StopPrimitive()
-                )
+            for robot_id, num_times_to_stop in self.robots_to_be_disconnected.items():
+                if num_times_to_stop > 0:
+                    robot_primitives[robot_id] = Primitive(stop=StopPrimitive())
+                    self.robots_to_be_disconnected[robot_id] = num_times_to_stop - 1
```

Comment:
if `num_times_to_stop` is 0, I think we should just remove that robot id from `robots_to_be_disconnected`

Reply:
done